src/game.cpp

```diff
@@ -1533,13 +1533,14 @@ bool game::do_turn()
 
             mon_info_update();
 
-            // If player is performing a task and a monster is dangerously close, warn them
-            // regardless of previous safemode warnings
+            // If player is performing a task, a monster is dangerously close, and monster can reach to the player,
+            // warn them regardless of previous safemode warnings
             if( u.activity && !u.has_activity( activity_id( "ACT_AIM" ) ) &&
                 u.activity.moves_left > 0 &&
                 !u.activity.is_distraction_ignored( distraction_type::hostile_spotted_near ) ) {
                 Creature *hostile_critter = is_hostile_very_close();
-                if( hostile_critter != nullptr ) {
+                if( hostile_critter != nullptr &&
+                    get_map().clear_path( hostile_critter->pos(), u.pos(), DANGEROUS_PROXIMITY, 1, 100 ) ) {
```

Comment:
I'm not sure about passing `cost_max=100` (last argument of the `clear_path`). 

Internally, `clear_path` checks this:
```cpp
const int cost = this->move_cost( new_point );
            if( cost < cost_min || cost > cost_max ) {
                is_clear = false;
                return false;
            }
```
If `move_cost` is 0 if the point is impassable. As you want to check that at least one tile has `move_cost==0`, you need to pass `cost_min=1` and `cost_max=INT_MAX`. 

Also, note that `clear_path` uses bresenham, so it only checks tiles that are directly on the line between two points. Which means, that if there is an impassable single-tile obstacle directly between player and monster, but monster can easily walk around it, this check will fail.
For example, if you modify your test a bit, I think you will not get a warning:
![image](https://user-images.githubusercontent.com/2865203/100790159-42ae5b80-33cc-11eb-8a0f-1dd5f2ad6172.png)

Ideally you'd want to use `map::route` instead of `clear_path`.

Reply:
Ok, I rework the logic to check for `route` instead of `clear_path`, but I'm not sure why it's working weird.

I'm checking whether monster has found the path to me, and if it's found, popup a debug message and trigger the prompt:

```
const std::vector<tripoint> path = get_map().route( hostile_critter->pos(), u.pos(),
                                                       hostile_critter->get_pathfinding_settings(), hostile_critter->get_path_avoid() );
if( !path.empty() ) {
 popup("path found");
 cancel_activity_or_ignore_query( distraction_type::hostile_spotted_near,
                                                   string_format( _( "The %s is dangerously close!" ),
                                                   hostile_critter->get_name() ) );
}
```
![изображение](https://user-images.githubusercontent.com/11132525/100848243-4a3c3600-349a-11eb-8404-184c6f2dd502.png)

I pressed enter to discard the popup, but for some reason the prompt isn't triggering on the very moment of time, but rather when the zombie is standing almost right next to me, 5 seconds later:
![изображение](https://user-images.githubusercontent.com/11132525/100848423-87082d00-349a-11eb-848a-8ce692e07bd3.png)

And if I chose "Yes", one more second passes, and the zombie makes one more step towards me:
![изображение](https://user-images.githubusercontent.com/11132525/100848629-d0f11300-349a-11eb-96c1-75fd565f40cf.png)

Reply:
Yea, that's weird 🤷‍♂️
Perhaps, try to see what's going on in debugger?

Reply:
Sadly I couldn't get debugger working despite my best efforts. Looks like I need a step-by-step manual on that.

Reply:
> Sadly I couldn't get debugger working despite my best efforts. Looks like I need a step-by-step manual on that.
I'm on mac and using clion. Let me know if I can be of any help.

Regarding other issues.

The "monster is spotted" notification is triggered from the `game::mon_info_update` method, that is invoked two times, before and after `monmove`. That means, that immediately after monster moves and becomes visible to the player, the "monster is spotted" notification is triggered.

But the "The monster is dangerously close!" notification is called **before** the monmove, which means that monsters do extra turn between monster is seen and the notification gets triggered.

Also, looking at how the `wait_redraw` is handled, seem like it could skip one frame immediately after activity is cancelled.

Unfortunately I cannot reproduce the weirdness between `popup` and `cancel_activity_or_ignore_query`. 

I believe the issues described above are outside of the scope of this PR and should be addressed separately.